include: zephyr: devicetree: add DT_INST_IRQN_BY_NAME()

danieldegrasse opened the pull request:
This macro should have been included with DT_INST_IRQN() (and is referenced in the migration guide), but was absent. Add the infrastructure to use it.

danieldegrasse committed:
include: zephyr: devicetree: add DT_INST_IRQN_BY_NAME()

This macro should have been included with DT_INST_IRQN() (and is
referenced in the migration guide), but was absent. Add the
infrastructure to use it.

Fixes #98858

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

rruuaanng reviewed (commented):
Please add some tests for the macro :)

danieldegrasse committed:
tests: lib: devicetree: add tests for DT_IRQN_BY_NAME

Add tests for DT_IRQN_BY_NAME macro to devicetree API test

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse commented:
> Please add some tests for the macro :)

Done, sorry it took me so long to return to this

mbolivar reviewed (commented):
LGTM, @kartben's comment about release notes should be addressed

danieldegrasse committed:
doc: releases: 4.5: add release note entry for DT_IRQN_BY_NAME

Add a release note entry for DT_IRQN_BY_NAME. This macro was previously
documented as being included in 3.6, but was not actually added to the
codebase.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

jhedberg commented:
> > @danieldegrasse - this is probably a bug and we should aim to fix it in 4.3.0
> 
> Updated- my assumption was that this would wait for 4.4, but I'm fine to fix it for this release if you want to

On the surface (since the subject starts with "add ...") it looks like a feature addition. Because of that it'd be imporatant to have a matching bug report issue which describes what's currently broken (which this PR then fixes).

rruuaanng commented on doc/releases/release-notes-4.3.rst:
```diff
@@ -169,6 +169,10 @@ New APIs and options
 
+* Devicetree
+
+  * :c:macro:`DT_IRQN_BY_NAME`
```
```suggestion
  * :c:macro:`DT_IRQN_BY_NAME`
  * :c:macro:`DT_INST_IRQN_BY_NAME`
```

mbolivar commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -873,6 +873,20 @@ ZTEST(devicetree_api, test_irq)
 #endif
 
+	/* DT_IRQN_BY_NAME */
+#ifndef CONFIG_MULTI_LEVEL_INTERRUPTS
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), err), 30, "");
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), stat), 40, "");
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), done), 60, "");
+#else
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), err),
+		      ((30 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 11, "");
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), stat),
+		      ((40 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 11, "");
+	zassert_equal(DT_IRQN_BY_NAME(DT_INST(0, DT_DRV_COMPAT), done),
+		      ((60 + 1) << CONFIG_1ST_LEVEL_INTERRUPT_BITS) | 11, "");
+#endif
```
If you're taking this route, I think it would be nicer to have the testcase.yaml run with both configurations to avoid regressions, but I guess it's not a blocker.

danieldegrasse replied:
`qemu_riscv32` is built with `CONFIG_MULTI_LEVEL_INTERRUPTS` and `native_sim` is built without it. Since both configurations can run in simulation I think we should be covered, yeah?

mbolivar reviewed (commented):
Looks OK, but needs update to 4.4

dkalowsk commented:
@danieldegrasse is this still needed?  4.4 development is open and the release-notes would need updating here.  Also does this need to be backported to 4.3 now as well?

rruuaanng reviewed (commented):
Move to `4.5`.

henrikbrixandersen commented:
@danieldegrasse Please rebase on `main`.

danieldegrasse commented:
Rebased to fix release notes conflict

mbolivar commented:
@danieldegrasse please fix the docs build, otherwise lgtm

danieldegrasse commented:
> @danieldegrasse please fix the docs build, otherwise lgtm

I think https://github.com/zephyrproject-rtos/zephyr/pull/109651 should fix that, I'll rekick CI and see if it works without a rebase
